### PR TITLE
Implement base solver

### PR DESCRIPTION
**Pull request by pierreguilmin**
This separates the API layer from the diffrax solvers implementation. It also allows implementing custom solvers very easily (not necessarily based on diffrax). I will use this to implement the propagator solver in a subsequent PR.

### CODE REVIEW

File: dynamiqs/solver.py [hunk 1]
```diff
@@ -1,27 +1,42 @@
...
 from jaxtyping import Scalar
 
 from .gradient import Adjoint, Autograd, Gradient
 
+NoneType = type(None)
+
 
 class Solver(eqx.Module):
-    SUPPORTED_GRADIENT: ClassVar[tuple[Gradient]] = ()
+    SUPPORTED_GRADIENT: ClassVar[tuple[Gradient | NoneType]] = (NoneType,)
```
Comment (abocquet) [inline dynamiqs/solver.py:14]: Is is true to say that all solvers can accept `None` as gradient? If so we could probably strip this `NoneType` type away

File: dynamiqs/core/diffrax_solver.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
+from __future__ import annotations
+
+import warnings
+from collections import namedtuple
+
+import diffrax as dx
+
+# from equinox import AbstractVar
```
Comment (abocquet) [inline dynamiqs/core/diffrax_solver.py:8]:
```suggestion
```

File: dynamiqs/solver.py (hunk 1 above)
Comment (pierreguilmin) [reply dynamiqs/solver.py:14]: It's true for now so let's remove it yes 👍🏼

File: dynamiqs/core/diffrax_solver.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+class DiffraxSolver(BaseSolver):
+    diffrax_solver: dx.AbstractSolver
+    stepsize_controller: dx.AbstractAdaptiveStepSizeController
+    dt0: float | None
+    max_steps: int
+    term: dx.ODETerm
+
+    def __init__(self, *args):
+        super().__init__(*args)
```
Comment (gautierronan) [inline dynamiqs/core/diffrax_solver.py:36]:
```suggestion
```

**Review by gautierronan — APPROVED** [review]
LGTM :)

File: dynamiqs/core/diffrax_solver.py (hunk 3 above)
Comment (pierreguilmin) [reply dynamiqs/core/diffrax_solver.py:36]: Fun fact, we need this 😅 
```python
import equinox as eqx

class A(eqx.Module):
    a: int

class B(A):
    b: int

#    def __init__(self, *args):
#        super().__init__(*args)

class C(B):
    def __init__(self, *args):
        super().__init__(*args)
        self.b = 2 * self.a

C(1)  # TypeError: __init__() missing 1 required positional argument: 'b'
```
If you uncomment the init of `B` it works as intended...

Maybe I'm missing smthg obvious here, cc @abocquet.

Comment (gautierronan) [reply dynamiqs/core/diffrax_solver.py:36]: Ok that's weird ! Not super important anyway, let's keep it ^^

**Pull request merged**